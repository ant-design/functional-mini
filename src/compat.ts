export * from './3rd-party/compat.js';

import compat from './3rd-party/compat.js';
export default compat;
