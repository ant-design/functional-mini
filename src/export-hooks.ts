import {
  useState,
  useEffect,
  useMemo,
  useCallback,
  useDebugValue,
  useRef,
} from './r.js';

import { useEvent } from './hooks.js';

export {
  useEvent,
  useCallback,
  useEffect,
  useMemo,
  useState,
  useDebugValue,
  useRef,
};
