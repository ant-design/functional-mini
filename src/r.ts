import { preactHooks, preact, serverRender } from './3rd-party/preact.js';
import { virtualDocument } from './3rd-party/virtual-document.js';
export { act } from './3rd-party/preact-test-utils.js';

export { serverRender };

const { createContext, h, render, options } = preact;

export { preactHooks };

export type VNode = preact.VNode;

export const React = {
  createElement: h,
  h: h,
  createContext,
  render,
  ...preactHooks,
};

options.requestAnimationFrame = (cb) => {
  return setTimeout(cb);
};

options.debounceRendering = (cb) => {
  return setTimeout(cb);
};

export const mountElement = function (element: any) {
  const entryDom = virtualDocument.createElement('div') as any;
  render(element, entryDom);

  return {
    unmount() {
      render(h('div', {}, 'unmounted'), entryDom);
    },

    //@ts-expect-error
    update(newElement) {
      render(newElement, entryDom);
    },

    toString() {
      return entryDom.innerHTML;
    },

    dom: entryDom,
  };
};
