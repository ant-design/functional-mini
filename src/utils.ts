let logSwitch = true;

export const setLogSwitch = (on: boolean) => {
  logSwitch = on;
};

//@ts-expect-error
export const log = (...args) => {
  if (!logSwitch) return;

  console.log.apply(console, args);
};

//@ts-expect-error
export const error = (...args) => {
  console.error.apply(console, args);
};

export const time = (label?: string) => {
  if (!logSwitch) return;

  console.time(label);
};

export const timeEnd = (label?: string) => {
  if (!logSwitch) return;
  console.timeEnd(label);
};

//@ts-expect-error
export const shallowCompare = (obj1, obj2) => {
  if (obj1 === obj2) return true;
  if (typeof obj1 !== typeof obj2) return false;
  if (typeof obj1 !== 'object' || typeof obj2 !== 'object') return false;
  if (obj1 === null || obj2 === null) return false;
  if (Array.isArray(obj1) || Array.isArray(obj2)) return false;
  if (Object.keys(obj1).length !== Object.keys(obj2).length) return false;
  for (const key in obj1) {
    if (obj1[key] !== obj2[key]) return false;
  }
  return true;
};

// 这么做是为了测试方便，隔离一些全局变量
export enum EComponent2Status {
  VALID = 'VALID',
  INVALID = 'INVALID',
  UNKNOWN = 'UNKNOWN',
}

let component2Status = EComponent2Status.UNKNOWN;
export const getComponent2Status = () => {
  return component2Status;
};

export const updateComponent2Status = (status?: EComponent2Status) => {
  if (status) {
    component2Status = status;
    //@ts-expect-error
  } else if (typeof my !== 'undefined') {
    //@ts-expect-error
    component2Status = my.canIUse('component2')
      ? EComponent2Status.VALID
      : EComponent2Status.INVALID;
  }
};

let incrementalId = 0;
const instanceKeyId = '_functional_instance_id_';

//@ts-expect-error
export function getIdFromAppxInstance(appxInstance) {
  if (!appxInstance) throw new Error('appxInstance param is falsy');
  let id;
  // 支付宝端
  if (appxInstance.$id) {
    id = `${appxInstance.$id}-${appxInstance.$page?.$viewId || ''}-${
      appxInstance.is || 'page'
    }`; /* 页面 */
  } else if (appxInstance[instanceKeyId]) {
    id = appxInstance[instanceKeyId];
  } else {
    id = `instance-${incrementalId++}`;
    appxInstance[instanceKeyId] = id;
  }
  if (!id) throw new Error('failed to identify appxInstance.$id');
  return id;
}

export const instanceKeyPropNames = '_functional_instance_prop_names_';

export function mergeObjectKeys(
  ...objects: Record<string, unknown>[]
): string[] {
  let res: string[] = [];
  for (const iterator of objects) {
    res = res.concat(Object.keys(iterator));
  }
  return res;
}
