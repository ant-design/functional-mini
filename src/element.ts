import { VNode, act, mountElement, serverRender, createElement } from './r.js';
import {
  log as globalLog,
  error as globalError,
  EComponent2Status,
  getComponent2Status,
  updateComponent2Status,
  getIdFromAppxInstance,
  instanceKeyPropNames,
  shallowCompare,
} from './utils.js';

import HandlersController from './handlers.js';
import { IElementContext, reactContext, useSyncMiniData } from './hooks.js';
import { ETargetPlatform, EElementType } from './types.js';
import { platformConfig } from './platform.js';

const FUNCTIONAL_MINI_PAGE_DOM_PLACEHOLDER = 'MINIFISH_PAGE_DOM_PLACEHOLDER';
const DANGER_ZONE_BYPASS_FUNCTION_CALL_WITH_DATA =
  'DANGER_ZONE_BYPASS_FUNCTION_CALL_WITH_DATA'; // 绕过所有函数方法，直接返回 data。不要在生产环境使用！

export interface IAppxOptions {
  data: Record<string, any>;
  options: Record<string, string>;
  methods?: Record<string, (args: any) => void>;
  [handlerName: string]: any;
}

export type TElementFunction<TProps> = (props: TProps) => {
  [axmlKey: string]: any;
};

export interface IInstanceMap {
  [id: string]: {
    appxContext: IElementContext;
    appxId: string;
    elementFn: any;
    elementInstance?: any;
    pendingProps: any;
    unmounted: boolean;
    propKeys: string[];
  };
}

function compositeElementWithContext(
  //@ts-expect-error
  id,
  //@ts-expect-error
  elementFn,
  appxContext: IElementContext,
  //@ts-expect-error
  pendingProps,
) {
  const el = createElement(elementFn, pendingProps);
  // eslint-disable-next-line react/no-children-prop
  return createElement(
    reactContext.Provider,
    { value: appxContext, key: id, children: [el] },
    el,
  ) as VNode;
}

export function flushReactTree(elementMap: IInstanceMap) {
  const children: VNode[] = [];
  for (const id in elementMap) {
    if (!Object.prototype.hasOwnProperty.call(elementMap, id)) continue;

    const item = elementMap[id];
    if (item.unmounted) {
      continue;
    }
    if (item.elementInstance && !item.pendingProps) {
      children.push(item.elementInstance);
      continue;
    }

    const contextElement = compositeElementWithContext(
      id,
      item.elementFn,
      item.appxContext,
      item.pendingProps,
    );
    children.push(contextElement);
    item.elementInstance = contextElement;
    item.pendingProps = null;
  }

  const parent = createElement('div', {}, children);
  return parent;
}

export interface ElementOption {
  options?: any;
}

export function functionalMiniElement<TProps>(
  element: TElementFunction<TProps>,
  displayName = '' /* 用于问题排查，和小程序 axml 无关 */,
  elementType: EElementType,
  defaultProps?: TProps,
  _targetPlatform?: ETargetPlatform,
  elementOption: ElementOption = {},
) {
  let targetPlatform: ETargetPlatform =
    _targetPlatform ?? ETargetPlatform.alipay;

  // 配置客户端环境
  const {
    pageEvents,
    componentEvents,
    buildOptions,
    componentLifeCycleToMount,
    componentLifeCycleToUnmount,
    pageLifeCycleToMount,
    pageLifeCycleToUnmount,
    getPropsFromInstance,
  } = platformConfig[targetPlatform];

  displayName = displayName || element.name;
  if (!displayName) {
    console.warn(
      '为了方便问题排查，请传入组件的 displayName 参数，或不要使用匿名函数组件 https://medium.com/@stevemao/do-not-use-anonymous-functions-to-construct-react-functional-components-c5408ec8f4c7',
    );
  }

  const nameTag = `[${elementType}/${displayName || '(unnamed)'}]`;
  //@ts-expect-error
  const log = (...args) => {
    globalLog.apply(null, [nameTag].concat(args));
  };
  const logErrorAndThrow = (err: Error) => {
    if (!err) return;
    err.message = `${nameTag} ${err.message}`;
    globalError(err);
    throw err;
  };

  if (elementType === 'component') {
    if (getComponent2Status() === EComponent2Status.UNKNOWN) {
      updateComponent2Status();
    }
    if (getComponent2Status() === EComponent2Status.INVALID) {
      // 如果还是 Unknown，就先不管了
      throw new Error(
        `无法注册 ${nameTag} 组件，因为当前小程序环境未开启 component2。配置入口在 mini.project.json : { "compileOptions": { "component2": true } }，或在 IDE 详情>项目配置中勾选`,
      );
    }
  }

  const platformExposedEvents: string[] =
    elementType === 'page' ? pageEvents : componentEvents;

  const defaultPropKeys = Object.keys(defaultProps || {});
  let observers: Record<string, unknown> = {};
  const elementMap: IInstanceMap = {};
  //@ts-expect-error
  let commonTestRenderer;
  function updateReactTree() {
    act(() => {
      const parent = flushReactTree(elementMap);
      //@ts-expect-error
      if (!commonTestRenderer) {
        commonTestRenderer = mountElement(parent);
      } else {
        commonTestRenderer.update(parent);
      }
    });
  }

  const WrappedElementFn = function (props: any) {
    let miniData: any;
    if (props && props[DANGER_ZONE_BYPASS_FUNCTION_CALL_WITH_DATA]) {
      miniData = props[DANGER_ZONE_BYPASS_FUNCTION_CALL_WITH_DATA];
      if (typeof miniData === 'string' && miniData.indexOf('%7B') === 0) {
        miniData = JSON.parse(decodeURIComponent(miniData));
      }
    } else {
      try {
        miniData = element.call(undefined, props);
      } catch (e) {
        //@ts-expect-error
        e.message = `渲染出错 ${e.message}`;
        return logErrorAndThrow(e as Error);
      }
    }
    if (typeof miniData === 'undefined') {
      log('函数组件没有返回渲染数据，请检查代码逻辑');
      miniData = {};
    } else if (typeof miniData !== 'object') {
      const e = new Error(
        `函数组件返回的渲染数据不合法，收到的类型为 ${typeof miniData}`,
      );
      return logErrorAndThrow(e);
    }
    // 不允许和 props 里的 key 重复
    const conflictKeys: string[] = [];
    const dataKeys = Object.keys(miniData) || [];
    for (const key in props) {
      if (dataKeys.indexOf(key) >= 0) {
        conflictKeys.push(key);
      }
    }
    if (conflictKeys.length > 0) {
      const e = new Error(
        `小程序自定义组件返回的渲染数据和 props 里的 key 重复：${conflictKeys.join(
          ', ',
        )}`,
      );
      return logErrorAndThrow(e);
    }

    // 上面的 early-return 都是检查和抛错，忽略 hooks 规则
    useSyncMiniData(miniData || {});
    return createElement('div', {}, FUNCTIONAL_MINI_PAGE_DOM_PLACEHOLDER);
  };

  // 在 onload 的时候，正式创建一个 react 组件
  function hookLoadToMountReactComponent(this: any, ...args: any[]) {
    const appxInstance = this;
    const id = getIdFromAppxInstance(appxInstance);
    if (elementMap[id])
      throw new Error(
        `duplicate id of appx instance, this might be a bug of minifish hooks. id: ${id}`,
      );
    const context = generateInstanceContext(appxInstance, false);
    appxInstance[instanceKeyPropNames] = defaultPropKeys;
    log('will mount react component');
    let initProps = {};
    if (elementType === 'component') {
      initProps = getPropsFromInstance(appxInstance, defaultPropKeys);
    }
    if (elementType === 'page') {
      initProps = {
        query: args[0] || {},
      };
    }

    elementMap[id] = {
      appxContext: context,
      appxId: id,
      elementFn: WrappedElementFn,
      pendingProps: initProps,
      unmounted: false,
      propKeys: defaultPropKeys,
    };
    updateReactTree();

    // 触发 onLoad
    if (elementType === 'page') {
      log('页面（Page）已经 Mount，开始触发 onLoad');
      return context.handlersController.callHandlers(
        pageLifeCycleToMount,
        appxInstance,
        args,
      ); // 直接调用controller 内部的方法，插队执行 onLoad
    } else {
      log(`组件（Component）已经 Mount，开始触发 ${componentLifeCycleToMount}`);
      return context.handlersController.callHandlers(
        componentLifeCycleToMount,
        appxInstance,
        args,
      );
    }
  }

  //@ts-expect-error
  function dispatchNewProps(appxInstance, nextProps) {
    const id = getIdFromAppxInstance(appxInstance);
    const instance = elementMap[id];
    if (instance) {
      instance.pendingProps = nextProps;
      updateReactTree();
    }
  }

  function hookUnloadToUnmount(this: any) {
    const appxInstance = this;
    const id = getIdFromAppxInstance(appxInstance);
    log(`will unmount react element of ${id}`);
    if (!elementMap[id]) {
      log(
        `找不到 id 为 ${id} 的实例，请检查是否开启了 component2 。这也可能是 Minifish 的 bug。`,
      );
    }
    elementMap[id].unmounted = true;
    updateReactTree();
    // TODO: 清理 appx context
  }

  const handlersController = new HandlersController(nameTag);
  //@ts-expect-error
  const anyUnknownContext = (ctx) => {
    if (!ctx) throw new Error('ctx is required');
    const id = getIdFromAppxInstance(ctx);
    return !elementMap[id];
  };
  const anyContext = () => true;
  if (elementType === 'page') {
    handlersController.addHandler(
      pageLifeCycleToMount,
      anyUnknownContext,
      hookLoadToMountReactComponent,
    );
    handlersController.addHandler(
      pageLifeCycleToUnmount,
      anyContext,
      hookUnloadToUnmount,
    );
  } else {
    // 组件的关键生命周期
    handlersController.addHandler(
      componentLifeCycleToMount,
      anyUnknownContext,
      hookLoadToMountReactComponent,
    );
    handlersController.addHandler(
      componentLifeCycleToUnmount,
      anyContext,
      hookUnloadToUnmount,
    );
    // 把 props 变更分发到对应的组件里
    if (targetPlatform === ETargetPlatform.alipay) {
      handlersController.addHandler(
        'deriveDataFromProps',
        anyContext,
        function hookDeriveDataFromProps(this: any, nextProps) {
          if (this.props && this.props === nextProps) {
            // 可能是setData触发的，不要死循环了
            return;
          } else {
            if (shallowCompare(this.props, nextProps, ['$slots'])) {
              return;
            }
          }
          return dispatchNewProps(this, nextProps);
        },
      );
    } else if (targetPlatform === ETargetPlatform.wechat) {
      const props = defaultPropKeys.join(', ');
      log(`将注册以下 key 的 props 更新：${props}`);
      observers = Object.assign(
        {
          // 忽略函数参数，直接从 this 里面找
          // 搞不懂小程序的设计逻辑，为什么非要在 props 里夹着 data
          [props](this: any, ...args: any[]) {
            log('observer is being called', args);
            const newProps = getPropsFromInstance(this, defaultPropKeys);
            return dispatchNewProps(this, newProps);
            //
          },
        },
        observers,
      );
    }
  }

  // 做一次预渲染，获取所有 appx 需要的属性
  const generateInstanceContext = (
    //@ts-expect-error
    instance,
    ifServerRender: boolean,
  ): IElementContext => {
    return {
      instance,
      handlersController,
      ifServerRender,
      debugLog: log,
      platformConfig: platformConfig[targetPlatform],
    };
  };

  // 收集 initData
  let initData: unknown = {};
  const fakeAppxInstance = {
    $id: `_minifish_hooks_pre_render_${Math.random()}`,
    //@ts-expect-error
    setData(data) {
      initData = Object.assign({}, initData || {}, data);
    },
  };
  const fakeCtx = generateInstanceContext(fakeAppxInstance, true);
  const serverEl = compositeElementWithContext(
    fakeAppxInstance.$id,
    WrappedElementFn,
    fakeCtx,
    defaultProps || {},
  );
  serverRender(serverEl);
  log('serverRendered with initData', initData);

  // 拼装对象，喂给 appx。产出的配置应该只能喂一次，喂多了 appx 不认。
  handlersController.lockHandlerNames();

  // 把生命周期和一般事件处理分开归类，因为组件的事件处理要多裹一层
  const lifeCycleHandlers = {};
  const userEventHandlers = {};
  const handlers = handlersController.getHandlersImplProxy();
  for (const name in handlers) {
    if (platformExposedEvents.indexOf(name) >= 0) {
      //@ts-expect-error
      lifeCycleHandlers[name] = handlers[name];
    } else {
      //@ts-expect-error
      userEventHandlers[name] = handlers[name];
    }
  }

  const finalOptions = buildOptions(
    elementType,
    defaultProps,
    initData,
    lifeCycleHandlers,
    userEventHandlers,
    elementOption.options,
    observers,
  );
  log('element options', finalOptions);
  return finalOptions;
}
