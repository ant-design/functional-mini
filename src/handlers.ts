import { log, error, time, timeEnd } from './utils.js';

export type TFunctionImpl = (...args: any[]) => any;

interface IImplInfo {
  fn: TFunctionImpl;
  bindContext: any;
  off?: () => void; // 对应的反注册方法
}

/*
管理所有事件处理函数，并生成一个代理函数，允许用户替换其中的某些实现
受制于小程序的运行时机制，在 locked 之后，不允许变更 handlerNames
*/
export default class HandlersController {
  elementTag = ''; // 主要是 debug 使用
  locked = false; // 是否已经锁定，锁定之后不允许新增 handler
  executionInProgress = false;
  executionQueue: Array<() => void> = [];
  handlerNames: string[] = [];
  handlerImpl: Record<string, IImplInfo[]> = {};

  constructor(elementTag = '') {
    this.elementTag = elementTag;
  }

  addHandler(
    name: string,
    bindContext?: any,
    impl?: TFunctionImpl,
    disableMultiImpl?: boolean,
  ): () => void {
    if (this.locked && this.handlerNames.indexOf(name) < 0) {
      throw new Error(`不允许动态新增 handler: ${name}`);
    }

    if (impl && !bindContext) {
      throw new Error(`bindContext 不能为空，运行时需要用到它`);
    }

    if (bindContext && !impl) {
      throw new Error('missing impl');
    }

    if (typeof name !== 'string') {
      throw new Error(
        `name 的类型不合法（${typeof name}），它应该是一个字符串`,
      );
    } else if (typeof impl !== 'function' && typeof impl !== 'undefined') {
      throw new Error(`impl 的类型不合法（${typeof impl}），它应该是一个函数`);
    }

    if (this.handlerNames.indexOf(name) < 0) {
      this.handlerNames.push(name);
    }
    this.handlerImpl[name] = this.handlerImpl[name] || [];

    const self = this;
    if (!impl) {
      return () => {};
    }

    if (
      disableMultiImpl &&
      this.getHandlersByName(name, bindContext).length > 0
    ) {
      throw new Error(`${name} 已经注册了一个实现，不能再新增`);
    }
    const implInfo: IImplInfo = {
      // 裹一层，把 off 方法收集下来
      fn(this: any, ...args) {
        if (!this)
          throw new Error('missing context, cannot find matching handler');
        const ctx = self.filterContext(bindContext, this);
        // context 不匹配，不处理
        if (!ctx) throw new Error('handler context mismatch, refuse executing');
        try {
          const ret = impl.apply(this, args);
          if (typeof ret === 'function') {
            implInfo.off = ret; // 这个 off 其实已经不用了
          } else {
            return ret; // 如果不是 function，认为这个返回值有用，不是 off
          }
        } catch (e) {
          error(
            `${self.elementTag} ${name} 执行出错，错误信息：${
              (e as Error).message
            }`,
          );
          throw e;
        }
      },
      bindContext,
    };

    this.handlerImpl[name].push(implInfo);

    return () => {
      log(`off ${name} - ${this.elementTag}`);
      const handlers = this.handlerImpl[name]; // 这里闭包里有明确的 implInfo, 不需要根据 context 过滤

      if (!handlers) {
        const msg = `没有找到 ${name} 对应的实现，是否已经被 reset 了 ？请检查生命周期流程`;
        error(msg);
        throw new Error(msg);
      }
      const index = handlers.indexOf(implInfo);
      const offFunction = handlers[index].off;
      if (offFunction) {
        try {
          offFunction.call(undefined);
        } catch (e) {
          error(`执行 ${name} 的反注册方法时出错：`);
          error(e);
          throw e;
        }
      }
      if (index >= 0) {
        handlers.splice(index, 1);
      }
    };
  }

  lockHandlerNames() {
    this.locked = true;
  }

  resetAllImpl() {
    this.handlerImpl = {};
  }

  getHandlersByName(name: string, context: any): IImplInfo[] {
    if (!context) throw new Error('context is required');
    return this.handlerImpl[name].filter((handler) => {
      return this.filterContext(handler.bindContext, context);
    });
  }

  callHandlers(name: string, context: any, args: any[]) {
    const controller = this;
    const handlers = (controller.handlerImpl[name] || []).filter((handler) => {
      const targetCtx = controller.filterContext(handler?.bindContext, context);
      return !!targetCtx;
    }); // 过滤并复制出来，防止执行过程中有人往里面塞东西造成乱序
    const callId = `${controller.elementTag}-handlerCalled-${name}-${Math.floor(
      Math.random() * 1000,
    )}`;

    const tipText = `${callId} - handlers count ${handlers.length}`;
    log(Date.now(), tipText);
    let returnValue: any;
    for (let i = 0; i < handlers.length; i++) {
      log(Date.now(), `${tipText} - executing ${i + 1}/${handlers.length}`);
      time(callId);
      const handler = handlers[i];
      const ret = handler.fn.apply(context, args);
      if (i === 0) {
        // 只返回第一个
        returnValue = ret;
      }
      timeEnd(callId);
    }

    return returnValue;
  }

  // 返回给 appx 的可能是一个 Promise，无论 appx 等不等，这里都会串行执行
  getHandlersImplProxy() {
    if (!this.locked) {
      throw new Error('please lock the handler names first');
    }

    const handlers = this.handlerNames.reduce((acc, handlerName) => {
      const controller = this;
      //@ts-expect-error
      acc[handlerName] = function (this: any, ...args) {
        // 这里要保留调用方的 this
        const appxCaller = this;
        if (!appxCaller) {
          console.warn('cannot find appx caller instance');
        }
        // deriveData 的时候，异步可能会有问题，直接改同步看看疗效
        let err;
        let ret;
        try {
          ret = controller.callHandlers(handlerName, appxCaller, args);
        } catch (e) {
          err = e;
        }

        // await unlock();
        // 防止后续所有事情被死锁
        if (err) throw err;

        // setTimeout(() => {
        // unlock();
        // }, 0);

        return ret;
        // return Promise.resolve(
        //   (async () => {

        //   })(),
        // );
      };
      return acc;
    }, {});

    return handlers;
  }

  //@ts-expect-error
  private filterContext(bindContext, nowContext) {
    if (typeof bindContext === 'function') {
      if (!bindContext(nowContext)) {
        return false;
      }
    } else if (bindContext !== nowContext) {
      return false;
    }
    return nowContext;
  }
}
