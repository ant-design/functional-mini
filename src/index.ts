import { setLogSwitch } from './utils.js';

export { setLogSwitch };
