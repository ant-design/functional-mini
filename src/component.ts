import { functionalMiniElement, TElementFunction } from './element.js';
import { getLifeCycleHooks } from './hooks.js';
import { alipayComponentEvents, commonComponentEvents } from './platform.js';
import { EElementType, ETargetPlatform } from './types.js';

export function alipayComponent<TProps extends Record<string, any>>(
  element: TElementFunction<TProps>,
  defaultProps?: TProps,
) {
  return functionalMiniElement(
    element,
    '',
    EElementType.component,
    defaultProps,
    ETargetPlatform.alipay,
  );
}

export function wechatComponent<TProps extends Record<string, any>>(
  element: TElementFunction<TProps>,
  defaultProps?: TProps,
) {
  return functionalMiniElement(
    element,
    '',
    EElementType.component,
    defaultProps,
    ETargetPlatform.wechat,
  );
}

export const useAttached = getLifeCycleHooks(commonComponentEvents.attached);
export const useReady = getLifeCycleHooks(commonComponentEvents.ready);
export const useMoved = getLifeCycleHooks(commonComponentEvents.moved);
export const useDetached = getLifeCycleHooks(commonComponentEvents.detached);

export const useDidMount = getLifeCycleHooks(
  alipayComponentEvents.didMount,
  undefined,
  ETargetPlatform.alipay,
);

// onInit, deriveDataFromProps, didUnmount, didUpdate 是不开放 Hooks 的，请使用 React 生命周期方法

export * from './export-hooks.js';

export { useComponent } from './hooks.js';
