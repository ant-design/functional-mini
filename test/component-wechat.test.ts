/* eslint-disable no-prototype-builtins */
import { afterEach, beforeEach, describe, expect, test, vi } from 'vitest';
import {
  alipayComponent,
  useComponent,
  useEvent,
  wechatComponent,
  useAttached,
  usePageShow,
} from '../src/component';
import { useEffect, useState } from '../src/r';
import { mountWechatComponent, setupWechatEnv } from './utils/common';
import { delay } from './utils/utils';

describe('component - wechat', async () => {
  let off: () => void;
  beforeEach(() => {
    off = setupWechatEnv();
  });

  afterEach(() => {
    off();
  });

  test('basic opt', async () => {
    const C = function () {
      return { foo: 'long-content' };
    };
    const opt = wechatComponent(C, {
      ss: 'ssabc',
      value: 1234,
      bbb: true,
      aaa: ['1', 2, true],
      oo: {},
    });
    const optStr = JSON.stringify(
      opt,
      (key, value) => {
        if (typeof value === 'function') {
          return `[function ${value.name}]`;
        } else {
          return value;
        }
      },
      2,
    );
    expect(optStr).toMatchSnapshot();
  });

  test('invalid props type', async () => {
    const C = function () {
      return { foo: 'long-content' };
    };
    expect(() => {
      wechatComponent(C, {
        ff: () => {},
      });
    }).toThrow(/类型/);
  });

  test('init props 更新是有顺序的', async () => {
    const logProps = vi.fn();
    const C = function (props) {
      logProps(props);
      return { counterX2: props.counter * 2 };
    };
    const opt = wechatComponent(C, { counter: 10 });
    // 初始化 SSR 的数据
    expect(logProps).toBeCalledTimes(1);
    expect(opt.data).toEqual({ counterX2: 20 });

    // 加载之后，先用 initProps，再用真的初始化数据
    const instance = mountWechatComponent(opt, { counter: 100 });
    expect(instance.data).toEqual({ counterX2: 200, counter: 100 });
    expect(logProps).toBeCalledTimes(2);
    expect(logProps.mock.calls[1][0]).toEqual({ counter: 100 });
  });

  test('update props', async () => {
    const C = function (props) {
      return { data: 'aaa' };
    };
    const instance = mountWechatComponent(wechatComponent(C, { foo: 'bar' }));
    expect(instance.data).toEqual({ data: 'aaa', foo: 'bar' });

    instance.updateProps({ foo: 'bar2' });
    expect(instance.data).toEqual({ data: 'aaa', foo: 'bar2' });

    instance.updateProps({ data: 'cccc' }); // 其实不会生效
    expect(instance.data).toEqual({ data: 'aaa', foo: 'bar2' });

    instance.updateProps({ foo: 234 });
    expect(instance.data).toEqual({ data: 'aaa', foo: '' });
  });

  test('测试 handleResult', async () => {
    const C = function (props) {
      const [counter, setCounter] = useState(0);
      useEvent(
        'one',
        () => {
          return {
            counter,
            props: props,
          };
        },
        { handleResult: true },
      );
      useEvent('updateCount', (v) => {
        setCounter(v);
      });
      return { two: '2', props, counter };
    };
    const componentOptions = wechatComponent(C, { foo: 'bar' });
    expect(typeof componentOptions.methods.updateCount).toBe('function');
    expect(typeof componentOptions.methods.one).toBe('undefined');
    const instance = mountWechatComponent(componentOptions);
    expect(instance.data.one()).toEqual({
      counter: 0,
      props: {
        foo: 'bar',
      },
    });
    expect(instance.data.props).toEqual({
      foo: 'bar',
    });
    instance.updateProps({ foo: '2' });
    await instance.callMethod('updateCount', 3);
    await delay(10);
    expect(instance.data.one()).toEqual({
      counter: 3,
      props: {
        foo: '2',
      },
    });
    expect(instance.data.props).toEqual({
      foo: '2',
    });
  });

  test('测试 useEvent 和 properties 冲突', async () => {
    const C = function (props) {
      useEvent(
        'onFormat',
        () => {
          return props.onFormat();
        },
        { handleResult: true },
      );

      return {};
    };
    const componentOptions = wechatComponent(C, { onFormat: '' });
    expect(() => mountWechatComponent(componentOptions)).toThrow(
      '事件 onFormat 注册失败，在 handleResult 开启后，事件不能同时在 properties 与 useEvent 中定义。',
    );
  });

  test('测试 handleResult', async () => {
    const C = function (props) {
      useEvent(
        'one',
        () => {
          return 'one';
        },
        { handleResult: true },
      );

      return { ...props.data };
    };
    const componentOptions = wechatComponent(C, { data: { foo: 'bar' } });
    const instance = mountWechatComponent(componentOptions);
    expect(() =>
      instance.updateProps({
        data: { one: '2' },
      }),
    ).toThrow('one - 禁止修改 data 上已经存在的函数');
  });

  test('lifetimes', async () => {
    const createFn = vi.fn();
    const useAttachedFn = vi.fn();
    const C = function (props) {
      useEffect(createFn, []);
      useAttached(useAttachedFn, []);
      return {};
    };

    const instance = mountWechatComponent(wechatComponent(C, {}));
    expect(createFn).toBeCalledTimes(1);
    expect(useAttachedFn).toBeCalledTimes(1);

    const paramError = new Error('sample_error');
    instance.callLifecycle('error', paramError);
    await delay(5);
  });

  test('event trigger', async () => {
    const onEvt = vi.fn();
    const payload = { a: 1 };
    const evtOpt = { opt: 'aaaa' };
    const C = function (props) {
      const component = useComponent();
      useEvent(
        'clickSomething',
        () => {
          component.triggerEvent('clickSomething', payload, evtOpt);
        },
        [component],
      );

      return {};
    };
    const instance = mountWechatComponent(wechatComponent(C, {}), {}, onEvt);
    await delay(5);
    instance.callMethod('clickSomething');
    expect(onEvt).toBeCalledTimes(1);
    expect(onEvt.mock.calls[0]).toEqual(['clickSomething', payload, evtOpt]);
  });

  test('event trigger', async () => {
    const option = wechatComponent(
      () => {
        return {};
      },
      {
        value: null,
      },
    );
    expect(option.properties.value).toMatchInlineSnapshot(
      `
      {
        "type": null,
        "value": null,
      }
    `,
    );
  });
  test('微信测试 callPageLifecycle', async () => {
    const C = function () {
      usePageShow(() => {
        return 'pageshow';
      });
      return {};
    };
    const componentOptions = wechatComponent(C);
    const instance = mountWechatComponent(componentOptions);
    expect(typeof componentOptions.pageLifetimes.show).toBe('function');
    expect(instance.callPageLifecycle('show')).toEqual('pageshow');
  });

  test('test component option', async () => {
    const option = wechatComponent(
      () => {
        return {};
      },
      {},
      {
        options: {
          styleIsolation: 'shared',
        },
      },
    );
    expect(option.options).toMatchInlineSnapshot(`
      {
        "styleIsolation": "shared",
      }
    `);

    const alipayOption = alipayComponent(
      () => {
        return {};
      },
      {},
      {
        options: {
          styleIsolation: 'shared',
        },
      },
    );
    expect(alipayOption.options).toMatchInlineSnapshot(`
      {
        "lifetimes": true,
        "styleIsolation": "shared",
      }
    `);
  });
});
