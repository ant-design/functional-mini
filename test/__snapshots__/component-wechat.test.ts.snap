// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`component - wechat > basic opt 1`] = `
"{
  \\"pageLifetimes\\": {},
  \\"properties\\": {
    \\"ss\\": {
      \\"type\\": \\"[function String]\\",
      \\"value\\": \\"ssabc\\"
    },
    \\"value\\": {
      \\"type\\": \\"[function Number]\\",
      \\"value\\": 1234
    },
    \\"bbb\\": {
      \\"type\\": \\"[function Boolean]\\",
      \\"value\\": true
    },
    \\"aaa\\": {
      \\"type\\": \\"[function Array]\\",
      \\"value\\": [
        \\"1\\",
        2,
        true
      ]
    },
    \\"oo\\": {
      \\"type\\": \\"[function Object]\\",
      \\"value\\": {}
    }
  },
  \\"data\\": {
    \\"foo\\": \\"long-content\\"
  },
  \\"options\\": {},
  \\"observers\\": {
    \\"ss, value, bbb, aaa, oo\\": \\"[function ss, value, bbb, aaa, oo]\\"
  },
  \\"lifetimes\\": {
    \\"attached\\": \\"[function ]\\",
    \\"detached\\": \\"[function ]\\"
  },
  \\"methods\\": {}
}"
`;
